syntax = "proto3";

package BattleProtocol;

enum BattleTurnInfoType {
    A_TO_B  = 0;
    B_TO_A  = 1;
}

message BattleTurnInfo {
    bytes               skillName       = 1;
    BattleTurnInfoType  type            = 2;
    int32               selfDeltaHP     = 3;
    int32               destDeltaHP     = 4;
    int32               selfBuffID      = 5;
    int32               selfBuffLast    = 6;
    int32               destBuffID      = 7;
    int32               destBuffLast    = 8;
};

message BattleOperationInfo {
    int32   isUserA     = 1;
    int32   skillIndex  = 2;
    bytes   userName    = 3;
};

message BattleFinishInfo {
    enum BattleResult {
        WIN     = 0;
        LOSE    = 1;
    }
    enum FinishMode {
        NORMAL                  = 0;
        OPPOSITE_DISCONNECTED   = 1;
        DRAW                    = 2;
    }
    BattleResult    result  = 1;
    FinishMode      mode    = 2;
};

enum BattleMode {
    EXP_BATTLE  = 0;
    DUEL_BATTLE = 1;
};

message BattleStartRequest {
    BattleMode  battleMode      = 1; 
    bytes       fromUser        = 2;
    int32       fromUserPkmId   = 3;
    bytes       destUser        = 4;
    int32       serverPkmId     = 5;
};

enum BattleStartStatus {
    SUCCESS         = 0;
    REFUSED         = 1;
    ALREADY_START   = 2;
    DEST_NOT_ONLINE = 3;
};

message BattleStartResponse {
    BattleStartStatus   status  = 1;
    int32               isUserA = 2;
    int32               urPkmId = 3;
    int32               taPkmId = 4;
};

message BattleInviteRequest {
    int32   battleMode      = 1;
    bytes   fromUser        = 2;
    int32   fromUserPkmId   = 3;
};

message BattleInviteResponse {
    BattleMode          battleMode      = 1;
    bytes               fromUser        = 2;
    bytes               destUser        = 3;
    int32               destUserPkmId   = 4;
    BattleStartStatus   flag            = 5;
};

message BattleGiveupInfo {
    bytes   userName = 1;
};

message TrophiePkmIdInfo {
    int32   pkmID_1     = 1;
    int32   pkmID_2     = 2;
    int32   pkmID_3     = 3;
};